### PR TITLE
Introduce External Metrics API

### PR DESCRIPTION
**Pull request by kawych**
**What this PR does / why we need it**:
Introduce External Metrics API

This PR introduces External Metrics API which has been proposed: xref https://github.com/kubernetes/community/pull/1802

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by kawych** [comment]
/uncc @mwielgus @MaciekPytel @DirectXMan12 
As this was FYI, not a review request.

**Comment by piosz** [comment]
LGTM, as long as the related proposal is approved.

Also, could you please split the PR into 2 commits:
- the code you introduced
- autogen changes?

**Comment by DirectXMan12** [comment]
cc @kubernetes/api-reviewers

**Comment by kawych** [comment]
@piosz 
PR split to two commits. Also added generated protobuf.

**Comment by kawych** [comment]
/cc @zmerlynn 
To approve changes in hack/

File: staging/src/k8s.io/metrics/pkg/apis/external_metrics/v1beta1/types.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// a metric value for some object
+type ExternalMetricValue struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// the name of the metric
+	MetricName string `json:"metricName" protobuf:"bytes,2,name=metricName"`
+
+	// labels set identifying the value within metric
```
Comment (thockin) [inline staging/src/k8s.io/metrics/pkg/apis/external_metrics/v1beta1/types.go]: This comment is not very helpful - I have no idea what to expect to see here.

In general we avoid maps, but I have no context to evaluate this field to tell whether it is OK or not.

File: staging/src/k8s.io/metrics/pkg/apis/external_metrics/v1beta1/types.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	// indicates the time at which the metrics were produced
+	Timestamp metav1.Time `json:"timestamp" protobuf:"bytes,4,name=timestamp"`
+
+	// indicates the window ([Timestamp-Window, Timestamp]) from
+	// which these metrics were calculated, when returning rate
+	// metrics calculated from cumulative metrics (or zero for
+	// non-calculated instantaneous metrics).
+	WindowSeconds *int64 `json:"window,omitempty" protobuf:"bytes,5,opt,name=window"`
+
+	// the value of the metric for this
```
Comment (thockin) [inline staging/src/k8s.io/metrics/pkg/apis/external_metrics/v1beta1/types.go]: comment ends abruptly - unfinished thought?

**Review by thockin — COMMENTED** [review]
Please keep in mind that API comments turn into autogen docs - they need to be complete and useful.

**Comment by kawych** [comment]
@thockin - Comments addressed, docs clarified
/assign @zmerlynn 
/assign @krousey 

Since I'm OOO for the whole next week, I'm handing over the review to @MaciekPytel

**Pull request merged**